diff --git a/mojo/extensions.bzl b/mojo/extensions.bzl
index e34b4da..884254a 100644
--- a/mojo/extensions.bzl
+++ b/mojo/extensions.bzl
@@ -29,17 +29,25 @@ _NULL_SHAS = {
 }
 
 def _mojo_toolchain_impl(rctx):
-    base_url = rctx.attr.base_url or "https://whl.modular.com/nightly/mojo-compiler"
-    rctx.download_and_extract(
-        url = "{}/mojo_compiler-{}-py3-none-{}.whl".format(
-            base_url,
-            rctx.attr.version,
-            _PLATFORM_MAPPINGS[rctx.attr.platform],
-        ),
-        sha256 = _KNOWN_SHAS.get(rctx.attr.version, _NULL_SHAS)[rctx.attr.platform],
-        type = "zip",
-        strip_prefix = "mojo_compiler-{}.data/platlib/modular".format(rctx.attr.version),
-    )
+    raw_base_url = rctx.attr.base_url
+
+    # Temp compatibility with old base URLs
+    if raw_base_url and raw_base_url.endswith("mojo-compiler"):
+        raw_base_url = raw_base_url.removesuffix("mojo-compiler")
+    for whl in "mojo_compiler", "mojo_compiler_mojo_libs":
+        platform = _PLATFORM_MAPPINGS[rctx.attr.platform] if whl == "mojo_compiler" else "any"
+        base_url = (raw_base_url or "https://whl.modular.com/nightly/") + whl.replace("_", "-")
+        rctx.download_and_extract(
+            url = "{}/{}-{}-py3-none-{}.whl".format(
+                base_url,
+                whl,
+                rctx.attr.version,
+                platform,
+            ),
+            sha256 = _KNOWN_SHAS.get(rctx.attr.version, _NULL_SHAS)[rctx.attr.platform],
+            type = "zip",
+            strip_prefix = "{}-{}.data/platlib/modular".format(whl, rctx.attr.version),
+        )
 
     rctx.template(
         "BUILD.bazel",
