diff --git a/cc/private/link/cc_linking_helper.bzl b/cc/private/link/cc_linking_helper.bzl
index 4e70262d9..ddccf8e13 100644
--- a/cc/private/link/cc_linking_helper.bzl
+++ b/cc/private/link/cc_linking_helper.bzl
@@ -298,7 +298,7 @@ def _create_dynamic_link_actions(
         so_interface, _ = _get_linked_artifact(actions, name, LINK_TARGET_TYPE.INTERFACE_DYNAMIC_LIBRARY, cc_toolchain, link_artifact_name_suffix)

         # TODO(b/28946988): Remove this hard-coded flag.
-        if not feature_configuration.is_enabled("targets_windows"):
+        if not feature_configuration.is_enabled("targets_windows") and not feature_configuration.is_enabled("set_soname"):
             link_action_kwargs["linkopts"].append("-Wl,-soname=" + _cc_internal.dynamic_library_soname(
                 actions,
                 linker_output.short_path,
