diff --git a/pycross/private/tools/uv_translator.py b/pycross/private/tools/uv_translator.py
index cdee9e9..7bf4fc2 100644
--- a/pycross/private/tools/uv_translator.py
+++ b/pycross/private/tools/uv_translator.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import re
 from collections import defaultdict
 from collections.abc import Callable
 from dataclasses import dataclass
@@ -26,6 +27,8 @@ from pycross.private.tools.lock_model import PackageKey
 from pycross.private.tools.lock_model import RawLockSet
 from pycross.private.tools.lock_model import RawPackage
 
+EDITABLE_PATTERN = re.compile("^ *-e +")
+
 
 class LockfileIncompatibleException(Exception):
     pass
@@ -39,6 +42,28 @@ class MismatchedVersionException(Exception):
     pass
 
 
+def get_default_dependencies(lock: Dict[str, Any]) -> List[Requirement]:
+    deps = lock.get("project", {}).get("dependencies", [])
+    return [Requirement(dep) for dep in deps]
+
+
+def get_optional_dependencies(lock: Dict[str, Any]) -> Dict[str, List[Requirement]]:
+    dep_groups = lock.get("project", {}).get("optional-dependencies", {})
+    return {group: [Requirement(dep) for dep in deps] for group, deps in dep_groups.items()}
+
+
+def get_development_dependencies(lock: Dict[str, Any]) -> Dict[str, List[Requirement]]:
+    dep_groups = lock.get("dependency-groups", {})
+    # backwards-compatiblity for https://github.com/astral-sh/uv/pull/8272
+    # see: https://docs.astral.sh/uv/concepts/projects/dependencies/#legacy-dev-dependencies
+    legacy_dev_deps = lock.get("tool", {}).get("uv", {}).get("dev-dependencies", [])
+    if legacy_dev_deps:
+        dev_deps = dep_groups.get("dev", []) + legacy_dev_deps
+        dep_groups["dev"] = list(set(dev_deps))
+
+    return {group: [Requirement(EDITABLE_PATTERN.sub("", dep)) for dep in deps] for group, deps in dep_groups.items()}
+
+
 def _print_warn(msg):
     print("WARNING:", msg)
 
@@ -160,7 +185,7 @@ def validate_uv_lockfile_version(lock_dict: Dict[str, Any]) -> None:
 
 
 def translate(
-    project_name: str,
+    project_dict: Dict[str, Any],
     packages_list: list[Dict[str, Any]],
     requires_python: SpecifierSet,
     default_group: bool,
@@ -172,16 +197,9 @@ def translate(
 ) -> RawLockSet:
     requirements: List[Requirement] = []
 
-    project_info = next(pkg for pkg in packages_list if pkg["name"] == project_name)
-    default_dependencies = [Requirement(dep["name"]) for dep in project_info.get("dependencies", {})]
-    optional_dependencies = {
-        group: [Requirement(dep["name"]) for dep in deps]
-        for group, deps in project_info.get("optional-dependencies", {}).items()
-    }
-    development_dependencies = {
-        group: [Requirement(dep["name"]) for dep in deps]
-        for group, deps in project_info.get("dependency-groups", {}).items()
-    }
+    default_dependencies = get_default_dependencies(project_dict)
+    optional_dependencies = get_optional_dependencies(project_dict)
+    development_dependencies = get_development_dependencies(project_dict)
 
     if default_group:
         requirements.extend(default_dependencies)
@@ -426,9 +444,8 @@ def main(args: Any) -> None:
     project_files = read_files(args.project_file, args.lock_file)
 
     project_dict = project_files.project_file
-    project_name = project_dict["project"]["name"]
-
     lock_dict = project_files.lock_file
+
     validate_lockfile_version(lock_dict)
 
     # backwards-compatiblity for https://github.com/astral-sh/uv/pull/5861
@@ -437,7 +454,7 @@ def main(args: Any) -> None:
     requires_python = SpecifierSet(lock_dict.get("requires-python", ""))
 
     lock_set = translate(
-        project_name,
+        project_dict,
         packages_list,
         requires_python,
         default_group=args.default_group,
